File: python/day10/day10.py
import os
import time


from copy import deepcopy


def main():
    # input
    print(os.getcwd())
    day = "10"
    year = "2022"
    debug = True
    input_file = f'../inputs/day{day}{"_debug" if debug else ""}.txt'
    print(input_file)
    part1, part2 = 0, 0
    with open(input_file) as f:
        lines = f.read().splitlines()

    start_time = time.time()
    commands = []
    register_values = [1]
    for l in lines:
        if l[0] == 'n':
            commands.append((0, 0))
        else:
            commands.append((1, int(l.split(' ')[1])))

    # print(commands)
    for (c, arg) in commands:
        last_value = register_values[-1]
        register_values.append(last_value)
        if c != 0:
            register_values.append(last_value+arg)

    for i in range(20, len(register_values), 40):
        part1 += register_values[i-1] * i

    # part2
    display = [['.' for _ in range(40)] for _ in range(6)]

    for i in range(240):
        row, col = i//40, i % 40
        if (abs(col-register_values[i]) <= 1):
            display[row][col] = '#'
    output = '\n'.join([''.join(row) for row in display])
    print(output)

    # output
    duration = int((time.time() - start_time) * 1000000)
    header = "#" * 21
    print(
        f"{header}\n* AoC {year} - Day {day} *\n{header}\n\nPart 1:\t{part1}\nPart 2:\t{part2}\nTime:\t{duration // 1000} ms")

    for r in register_values: print(r)
if __name__ == "__main__":
    main()

File: js/day10/day10.js
let { input } = require("./input");
const day = 10;
const year = 2022;
let part1 = 0;
let part2 = 0;

function range(start, end, step = 1) {
  let result = [];
  if (step > 0) {
    for (let i = start; i < end; i += step) result.push(i);
  }
  if (step < 0) {
    for (let i = start; i > end; i += step) result.push(i);
  }
  return result;
}

// input = input[0].trim();
let commands = [];
for (let l of input) {
  tmp = l.trim().split(" ");
  if (tmp.length == 1) {
    commands.push({ op: 0, arg: 0 });
  } else {
    commands.push({ op: 1, arg: parseInt(tmp[1], 10) });
  }
}
// part 1
let registerValues = [1];

for (let { op, arg } of commands) {
  lastValue = registerValues.slice(-1)[0];
  registerValues.push(lastValue);
  if (op != 0) {
    registerValues.push(lastValue + arg);
  }
}
// console.log(registerValues)
for (let i of range(20, registerValues.length, (step = 40))) {
  part1 += registerValues[i - 1] * i;
}
// part2
let height = 6,
  width = 40;
let display = Array.from(Array(height), () =>
  Array.from({ length: width }, () => ".")
);
for (let i=0;i<height*width;i++){
    let row = ~~(i/width), col = i%width;
    if (Math.abs(col-registerValues[i])<=1){
        display[row][col] = '#';
    }
}
let output = display.map((line) => line.join()).join('\n');
console.log(
  `Advent of Code ${year} - Day ${String(day).padStart(2, "0")} \n${"#".repeat(
    28
  )}\nPart 1:\t${part1}\nPart 2:\t${part2}`
);
console.log(output);